pkgs/by-name/ni/nixos-rebuild-ng/src/nixos_rebuild/__init__.py

```diff
@@ -0,0 +1,203 @@
+        case Action.SWITCH | Action.BOOT:
+            info("building the system configuration...")
+            if args.rollback:
+                if action in (Action.TEST, Action.BUILD):
+                    raise NotImplementedError(
+                        f"Rollback not implement yet for {action}"
+                    )
+                path_to_config = rollback(profile)
+            elif flake:
+                path_to_config = nixos_build_flake(
+                    "toplevel",
+                    flake,
+                    no_link=True,
+                    **flake_flags,
+                )
+            else:
+                path_to_config = nixos_build(
+                    "system",
+                    args.attr,
+                    args.file,
+                    no_out_link=True,
+                )
+            set_profile(profile, path_to_config)
+            switch_to_configuration(
+                path_to_config,
+                action,
+                install_bootloader=args.install_bootloader,
+            )
+        case Action.TEST | Action.BUILD | Action.DRY_BUILD | Action.DRY_ACTIVATE:
+            info("building the system configuration...")
+            dry_run = action == Action.DRY_BUILD
+            if args.rollback:
+                if action in (Action.TEST, Action.BUILD):
+                    raise NotImplementedError(
+                        f"Rollback not implement yet for {action}"
+                    )
+                path_to_config = rollback(profile)
+            elif flake:
+                path_to_config = nixos_build_flake(
+                    "toplevel",
+                    flake,
+                    keep_going=True,
+                    dry_run=dry_run,
+                    **flake_flags,
+                )
+            else:
+                path_to_config = nixos_build(
+                    "system",
+                    args.attr,
+                    args.file,
+                    keep_going=True,
+                    dry_run=dry_run,
+                )
+            if action in (Action.TEST, Action.DRY_ACTIVATE):
+                switch_to_configuration(path_to_config, action)
+        case Action.BUILD_VM | Action.BUILD_VM_WITH_BOOTLOADER:
+            info("building the system configuration...")
+            attr = "vm" if action == Action.BUILD_VM else "vmWithBootLoader"
+            if flake:
+                path_to_config = nixos_build_flake(
+                    attr,
+                    flake,
+                    keep_going=True,
+                    **flake_flags,
+                )
+            else:
+                path_to_config = nixos_build(
+                    attr,
+                    args.attr,
+                    args.file,
+                    keep_going=True,
+                )
+            vm_path = next(path_to_config.glob("bin/run-*-vm"))
+            print(f"\nDone. The virtual machine can be started by running {vm_path}")
```

Comment:
While this part has some duplication, the separation of the actions here tries to make the logic as clear as possible, and is designed to reduce as much as possible big `if...else`.